proposals/NNNN-resource-arrays.md

```diff
@@ -0,0 +1,258 @@
+this:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 105, i1 false)
+```
+https://godbolt.org/z/hx99s5dzP
+
+#### Unbounded-size Arrays and Multiple Dimensions
+
+For multi-dimensional arrays of resources only the first dimension can be
+unbounded:
+```
+RWBuffer<float> E[][4] : register(u5);
+```
+When initializing individual resources of an unbounded multi-dimensional array,
+the arguments of `@dx.op.createHandleFromBinding` again look like as if the
+array was flattened to a single dimension. For example the call to create
+resource handle `E[100][4]` from array `E` declared above will have index `409`
+= `5 + 4 x 100 + 4`:
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 409, i1 false),
+```
+https://godbolt.org/z/YrEnYhrEj
+
+Arrays with unbounded sizes in the other dimensions are not allowed, though the
+error message DXC reports does not clearly indicate this:
+
+```
+RWBuffer<float> F[4][];
+
+error: array has incomplete element type 'RWBuffer<float> []'
+```
+https://godbolt.org/z/osq7dzvdh
+
+### Resource Arrays at Local Scope
+
+#### Local Resource Arrays with Fixed Size
+
+Resources array with fixed size can be declared as local variables in a function.
+Resource handles for the individual resources in the array are uninitialized
+until they are assigned a value from an initialized resource. DXC does not
+report error when an uninitialized resource is accessed and instead creates a
+call to `dx.op.annotateHandle` with zero-initialized handle value.
+
+```
+RWBuffer<float> Buf;
+RWBuffer<float> Out;
+
+float foo() {
+  RWBuffer<float> LocalArray[3];
+  LocalArray[2] = Buf;
+  // LocalArray[0] is not initialized
+  return LocalArray[2][0] + LocalArray[0][0];
+}
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = foo();
+}
+```
+https://godbolt.org/z/Gdc3q674W
+
+#### Unbounded Local Resource Arrays
+
+DXC even allows unbounded resource arrays as local variables in a function. This
+seems wrong from a language point of view. How is the compiler supposed to
+allocate such variable? 
+
+And as in the previous example, DXC does not report error when an uninitialized
+resource in an unbounded array is accessed and instead creates a call to
+`dx.op.annotateHandle` with zero-initialized handle value.
+
+```
+RWBuffer<float> Buf;
+RWBuffer<float> Out;
+
+float foo() {
+  RWBuffer<float> Unbounded[];
+  Unbounded[100] = Buf;
+  // Unbounded[0] is not initialized
+  return Unbounded[100][0] + Unbounded[0][0];
+}
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = foo();
+}
+```
+https://godbolt.org/z/43rGnY4Ee
+
+#### Resource Arrays as Function Arguments
+
+Both fixed-size and unbounded resource arrays can be used as function arguments.
+The dimensions of array passed into a function must match exactly the dimension
+of the declared argument
```

Comment:
I don't think we should continue to allow unbounded resource arrays as function arguments, unless there's a compelling use case.  If we did, I think they might be special, like only accepting another unbounded array, and requiring const inout (or const reference syntax when we have that).  This should restrict them to only ever coming from an unbounded global resource array, and we can then map them directly, without worrying about copy-in/copy-out.